File: rust/cudf-sys/cpp/sorting.cpp
// SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION.
// SPDX-License-Identifier: Apache-2.0

#include "cudf-sys/src/sorting.rs.h"
#include "cudf-sys/src/lib.rs.h"
#include "cudf-sys/helpers.hpp"

#include <cudf/sorting.hpp>

namespace cudf_sys {

std::unique_ptr<Table> sort_table(
    Table const& tbl,
    rust::Slice<int32_t const> column_orders,
    rust::Slice<int32_t const> null_orders,
    std::size_t stream) {
  return TBL(cudf::sort(tbl.cached_view(), ORDERS(column_orders), NULL_ORDERS(null_orders), S(stream)));
}

std::unique_ptr<Column> sorted_order(
    Table const& tbl,
    rust::Slice<int32_t const> column_orders,
    rust::Slice<int32_t const> null_orders,
    std::size_t stream) {
  return COL(cudf::sorted_order(tbl.cached_view(), ORDERS(column_orders), NULL_ORDERS(null_orders), S(stream)));
}

bool is_sorted_table(
    Table const& tbl,
    rust::Slice<int32_t const> column_orders,
    rust::Slice<int32_t const> null_orders,
    std::size_t stream) {
  return cudf::is_sorted(tbl.cached_view(), ORDERS(column_orders), NULL_ORDERS(null_orders), S(stream));
}

std::unique_ptr<Column> stable_sorted_order(
    Table const& tbl,
    rust::Slice<int32_t const> column_orders,
    rust::Slice<int32_t const> null_orders,
    std::size_t stream) {
  return COL(cudf::stable_sorted_order(tbl.cached_view(), ORDERS(column_orders), NULL_ORDERS(null_orders), S(stream)));
}

std::unique_ptr<Table> stable_sort_table(
    Table const& tbl,
    rust::Slice<int32_t const> column_orders,
    rust::Slice<int32_t const> null_orders,
    std::size_t stream) {
  return TBL(cudf::stable_sort(tbl.cached_view(), ORDERS(column_orders), NULL_ORDERS(null_orders), S(stream)));
}

std::unique_ptr<Table> sort_by_key(
    Table const& values,
    Table const& keys,
    rust::Slice<int32_t const> column_orders,
    rust::Slice<int32_t const> null_orders,
    std::size_t stream) {
  return TBL(cudf::sort_by_key(values.cached_view(), keys.cached_view(), ORDERS(column_orders), NULL_ORDERS(null_orders), S(stream)));
}

std::unique_ptr<Table> stable_sort_by_key(
    Table const& values,
    Table const& keys,
    rust::Slice<int32_t const> column_orders,
    rust::Slice<int32_t const> null_orders,
    std::size_t stream) {
  return TBL(cudf::stable_sort_by_key(values.cached_view(), keys.cached_view(), ORDERS(column_orders), NULL_ORDERS(null_orders), S(stream)));
}

std::unique_ptr<Column> rank_column(
    cudf::column_view const& col,
    int32_t method,
    int32_t column_order,
    int32_t null_handling,
    int32_t null_precedence,
    bool percentage,
    std::size_t stream) {
  return COL(cudf::rank(col,
      ENUM<cudf::rank_method>(method),
      ENUM<cudf::order>(column_order),
      ENUM<cudf::null_policy>(null_handling),
      ENUM<cudf::null_order>(null_precedence),
      percentage, S(stream)));
}

std::unique_ptr<Column> top_k(cudf::column_view const& col, int32_t k, int32_t order, std::size_t stream) {
  return COL(cudf::top_k(col, k, ENUM<cudf::order>(order), S(stream)));
}

std::unique_ptr<Column> top_k_order(cudf::column_view const& col, int32_t k, int32_t order, std::size_t stream) {
  return COL(cudf::top_k_order(col, k, ENUM<cudf::order>(order), S(stream)));
}

std::unique_ptr<Column> segmented_sorted_order(
    Table const& tbl,
    cudf::column_view const& segment_offsets,
    rust::Slice<int32_t const> column_orders,
    rust::Slice<int32_t const> null_orders,
    std::size_t stream) {
  return COL(cudf::segmented_sorted_order(tbl.cached_view(), segment_offsets, ORDERS(column_orders), NULL_ORDERS(null_orders), S(stream)));
}

std::unique_ptr<Column> segmented_top_k(cudf::column_view const& col, cudf::column_view const& segment_offsets, int32_t k, int32_t order, std::size_t stream) {
  return COL(cudf::segmented_top_k(col, segment_offsets, k, ENUM<cudf::order>(order), S(stream)));
}

std::unique_ptr<Column> segmented_top_k_order(cudf::column_view const& col, cudf::column_view const& segment_offsets, int32_t k, int32_t order, std::size_t stream) {
  return COL(cudf::segmented_top_k_order(col, segment_offsets, k, ENUM<cudf::order>(order), S(stream)));
}

std::unique_ptr<Column> stable_segmented_sorted_order(
    Table const& tbl,
    cudf::column_view const& segment_offsets,
    rust::Slice<int32_t const> column_orders,
    rust::Slice<int32_t const> null_orders,
    std::size_t stream) {
  return COL(cudf::stable_segmented_sorted_order(
      tbl.cached_view(), segment_offsets,
      ORDERS(column_orders), NULL_ORDERS(null_orders), S(stream)));
}

std::unique_ptr<Table> stable_segmented_sort_by_key(
    Table const& values,
    Table const& keys,
    cudf::column_view const& segment_offsets,
    rust::Slice<int32_t const> column_orders,
    rust::Slice<int32_t const> null_orders,
    std::size_t stream) {
  return TBL(cudf::stable_segmented_sort_by_key(
      values.cached_view(), keys.cached_view(), segment_offsets,
      ORDERS(column_orders), NULL_ORDERS(null_orders), S(stream)));
}

std::unique_ptr<Table> segmented_sort_by_key(
    Table const& values,
    Table const& keys,
    cudf::column_view const& segment_offsets,
    rust::Slice<int32_t const> column_orders,
    rust::Slice<int32_t const> null_orders,
    std::size_t stream) {
  return TBL(cudf::segmented_sort_by_key(
      values.cached_view(), keys.cached_view(), segment_offsets,
      ORDERS(column_orders), NULL_ORDERS(null_orders), S(stream)));
}

}  // namespace cudf_sys

File: rust/cudf-sys/src/sorting.rs
// SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION.
// SPDX-License-Identifier: Apache-2.0

#[cxx::bridge(namespace = "cudf_sys")]
pub mod ffi {
    unsafe extern "C++" {
        include!("cudf-sys/lib.hpp");

        // Shared types from the core bridge.
        type Column = crate::ffi::Column;
        type Table = crate::ffi::Table;
        #[namespace = "cudf"]
        type column_view = crate::ffi::column_view;

        // -- Sorting --

        /// Sorts the rows of a table.
        fn sort_table(
            tbl: &Table,
            column_orders: &[i32],
            null_orders: &[i32],
            stream: usize,
        ) -> Result<UniquePtr<Table>>;

        /// Returns the row indices that would sort the table.
        fn sorted_order(
            tbl: &Table,
            column_orders: &[i32],
            null_orders: &[i32],
            stream: usize,
        ) -> Result<UniquePtr<Column>>;

        /// Checks whether the rows of a table are sorted.
        fn is_sorted_table(
            tbl: &Table,
            column_orders: &[i32],
            null_orders: &[i32],
            stream: usize,
        ) -> Result<bool>;

        // -- Sorting (new) --

        /// Returns the row indices that would produce a stable sort of the table.
        fn stable_sorted_order(
            tbl: &Table,
            column_orders: &[i32],
            null_orders: &[i32],
            stream: usize,
        ) -> Result<UniquePtr<Column>>;

        /// Stable-sorts the rows of a table.
        fn stable_sort_table(
            tbl: &Table,
            column_orders: &[i32],
            null_orders: &[i32],
            stream: usize,
        ) -> Result<UniquePtr<Table>>;

        /// Sorts a table of values by a table of keys.
        fn sort_by_key(
            values: &Table,
            keys: &Table,
            column_orders: &[i32],
            null_orders: &[i32],
            stream: usize,
        ) -> Result<UniquePtr<Table>>;

        /// Stable-sorts a table of values by a table of keys.
        fn stable_sort_by_key(
            values: &Table,
            keys: &Table,
            column_orders: &[i32],
            null_orders: &[i32],
            stream: usize,
        ) -> Result<UniquePtr<Table>>;

        /// Computes the rank of each element in a column.
        fn rank_column(
            col: &column_view,
            method: i32,
            column_order: i32,
            null_handling: i32,
            null_precedence: i32,
            percentage: bool,
            stream: usize,
        ) -> Result<UniquePtr<Column>>;

        /// Returns the top-k elements of a column.
        fn top_k(col: &column_view, k: i32, order: i32, stream: usize)
        -> Result<UniquePtr<Column>>;

        /// Returns the row indices of the top-k elements of a column.
        fn top_k_order(
            col: &column_view,
            k: i32,
            order: i32,
            stream: usize,
        ) -> Result<UniquePtr<Column>>;

        /// Returns row indices that would sort each segment of a table.
        fn segmented_sorted_order(
            tbl: &Table,
            segment_offsets: &column_view,
            column_orders: &[i32],
            null_orders: &[i32],
            stream: usize,
        ) -> Result<UniquePtr<Column>>;

        /// Returns the top-k elements within each segment of a column.
        fn segmented_top_k(
            col: &column_view,
            segment_offsets: &column_view,
            k: i32,
            order: i32,
            stream: usize,
        ) -> Result<UniquePtr<Column>>;

        /// Returns the row indices of the top-k elements within each segment.
        fn segmented_top_k_order(
            col: &column_view,
            segment_offsets: &column_view,
            k: i32,
            order: i32,
            stream: usize,
        ) -> Result<UniquePtr<Column>>;

        // -- Sorting: stable segmented --

        /// Returns row indices for a stable sort within each segment.
        fn stable_segmented_sorted_order(
            tbl: &Table,
            segment_offsets: &column_view,
            column_orders: &[i32],
            null_orders: &[i32],
            stream: usize,
        ) -> Result<UniquePtr<Column>>;

        /// Stable-sorts values by keys within each segment.
        fn stable_segmented_sort_by_key(
            values: &Table,
            keys: &Table,
            segment_offsets: &column_view,
            column_orders: &[i32],
            null_orders: &[i32],
            stream: usize,
        ) -> Result<UniquePtr<Table>>;

        // -- Sorting: segmented_sort_by_key (non-stable) --

        /// Sorts values by keys within each segment.
        fn segmented_sort_by_key(
            values: &Table,
            keys: &Table,
            segment_offsets: &column_view,
            column_orders: &[i32],
            null_orders: &[i32],
            stream: usize,
        ) -> Result<UniquePtr<Table>>;
    }
}
